File: RACDaux/src/ode_immune.c
/*
 #      ____  ___   __________
 #     / __ \/   | / ____/ __ \
 #    / /_/ / /| |/ /   / / / /
 #   / _, _/ ___ / /___/ /_/ /
 #  /_/ |_/_/  |_\____/_____/
 #
 #  Sean Wu & John M. Marshall
 #  December 2017
 #
 #  Pre-erythrocytic immunity model
*/

#include <R.h>
#include <math.h>

/*
  Initial parameters

  # age and heterogeneity
  a0: age-dependent biting heterogeneity
  rho age-dependent biting heterogeneity
  zeta: relative biting rate
  psi:

  # immunity reducing probability of infection
  durB: 1/decay rate
  uB: duration in which immunity is not boosted
  b0: probability of infection with no immunity
  b1 maximum relative reduction
  IB0: scale parameter
  kappaB: shape parameter

  # immunity reducing probability of detection
  durD: 1/decay rate
  uD: duration in which immunity is not boosted

  # immunity reducing probabiltiy of clinical disease
  durC: 1/decay rate
  uC: duration in which immunity is not boosted

  # biting
  epsilon0: mean EIR for adults
*/
static double     parms[15];

#define zeta       parms[0]
#define psi        parms[1]
#define a0         parms[2]
#define rho        parms[3]

#define durB       parms[4]
#define uB         parms[5]
#define durD       parms[6]
#define uD         parms[7]
#define durC       parms[8]
#define uC         parms[9]

#define b0         parms[10]
#define b1         parms[11]
#define IB0        parms[12]
#define kappaB     parms[13]
#define epsilon0   parms[14]

/* initializer */
void init_immune(void (* odeparms)(int *, double *))
{
    int N = 15;
    odeparms(&N, parms);
}

/* derivatives */
void derivs_immune(int *neq, double *t, double *y, double *ydot, double *yout, int*ip){

  double time = *t;
  double epsilon = epsilon0*zeta*(1 - rho*exp(-time/a0))*psi; /*EIR at age a*/
  // double b = b0*(b1 + ((1-b1)/(1 + pow((y[0]/IB0),kappaB)))); /*mosquito to human transmission efficiency*/
  double lambda = epsilon*b0*(b1 + ((1.0-b1)/(1.0 + pow((y[0]/IB0),kappaB)))); /*force of infection at age a*/

  ydot[0] = epsilon/(epsilon*uB + 1.0) - y[0]/durB;
  ydot[1] = lambda/(lambda*uD + 1.0) - y[1]/durD;
  ydot[2] = lambda/(lambda*uC + 1.0) - y[2]/durC;

}

File: RACDaux/R/ode_immune.R
###############################################################################
#       ____  ___   __________
#      / __ \/   | / ____/ __ \
#     / /_/ / /| |/ /   / / / /
#    / _, _/ ___ / /___/ /_/ /
#   /_/ |_/_/  |_\____/_____/
#
#   Sean Wu & John Marshall
#   December 2017
#   ODE Immune Model Wrappers
#
###############################################################################


###############################################################################
# Within-host Immune ODE Model
###############################################################################

#' Within-host Immune ODE Model
#'
#' write me
#'
#' @param time times to evaluate state variables at
#' @param theta named vector of parameters
#' @param state initial states
#' @param ... additional parameters passed to numerical integrator \code{\link[deSolve]{ode}}
#'
#' @export
immune_ode <- function(time, theta, state, ...){
  return(deSolve::ode(y=state,parms=theta,time=time,func="derivs_immune",dllname="RACDaux",initfunc="init_immune",...))
}

#' Within-host Pre-erythrocytic Immune ODE Model
#'
#' Calculate an individual's pre-erythrocytic immunity at age a for given EIR heterogeneities (biting rate & geographic location).
#' This function calls \code{\link{immune_ode}}
#'
#' @param a age of individual
#' @param zeta individual biting heterogeneity
#' @param psi relative biting rate (used to calculate psi(a))
#' @param theta vector of model parameters (see \code{\link{RACD_Parameters}})
#' @param ... additional parameters passed to numerical integrator \code{\link[deSolve]{ode}}
#'
#' @export
immune_initial <- function(a, zeta, psi, theta, ...){

  theta_immune = c("zeta"=zeta,"psi"=psi,theta[c("a0","rho","dB","uB","dID","uD","dC","uC","b0","b1","IB0","kappaB","epsilon0")])
  theta_immune[["dB"]] = theta_immune[["dB"]]/365
  theta_immune[["uB"]] = theta_immune[["uB"]]/365
  theta_immune[["dID"]] = theta_immune[["dID"]]/365
  theta_immune[["uD"]] = theta_immune[["uD"]]/365
  theta_immune[["dC"]] = theta_immune[["dC"]]/365
  theta_immune[["uC"]] = theta_immune[["uC"]]/365
  theta_immune[["epsilon0"]] = theta_immune[["epsilon0"]]*365

  state_immune = c("IB"=0,"ID"=0,"ICA"=0)

  time_immune = seq(from=0,to=a,by=a/1e3)

  immune = RACDaux::immune_ode(time = time_immune,theta = theta_immune,state = state_immune,...)
  return(c(
    IB = unname(immune[nrow(immune),"IB"]),
    ID = unname(immune[nrow(immune),"ID"]),
    ICA = unname(immune[nrow(immune),"ICA"])
  ))
}
